### PR TITLE
14893: Edit message arrow up

### PR DESCRIPTION
**Pull request by gedu**
### Details

Now `ReportScreen` is in charge to sort the messages and passing them down, this way the children receive the same data.

### Fixed Issues

$ https://github.com/Expensify/App/issues/14893
PROPOSAL: https://github.com/Expensify/App/issues/14893#issuecomment-1431569016

### Tests

1. Sign in (if you are already signed in, first sign out)
2. Open a chat with messages
3. Click the arrow up and should send you to the last message

- [ ] Verify that no errors appear in the JS console

### Offline tests

Same steps as before

### QA Steps

Same steps as before

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/1676818/219367272-c8eb7928-2af4-4909-a5a6-5e76a2f3c09a.mp4

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/1676818/219367800-97df2dfc-2a36-4324-9a3b-9a68e44b3d79.mp4

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/1676818/219367851-5738b190-63bd-4630-aace-9e54918fa897.mp4

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/1676818/219367869-9e051435-c811-41b0-86de-342853bb92ca.mp4

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/1676818/219367932-51380468-fee1-4bd3-b121-3dadcc82fb17.mp4

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/1676818/219367962-47d7fce1-b533-4063-825a-ebf5572dd7e0.mp4

</details>

### CODE REVIEW

File: src/pages/home/report/ReportActionsView.js
Comment (mountiny) [inline src/pages/home/report/ReportActionsView.js:135]: Is this required in here as well if we have added the logic to the ReportScreen?

Comment (gedu) [reply src/pages/home/report/ReportActionsView.js:135]: Yes, the `mostRecentIOUReportActionID` is used just in this component and sends it down, the `ReportScreen` now just is taking care of the sorting, the logic for `mostRecentIOUReportActionID` isn't in `ReportScreen`

Comment (gedu) [reply src/pages/home/report/ReportActionsView.js:135]: ohhhh got it, yes it is true, my bad, I miss it after some merge conflicts, updating PR

File: src/pages/home/ReportScreen.js
Comment (mountiny) [inline src/pages/home/ReportScreen.js:135]: @gedu I am not very happy with moving all of this logic up which mens amore components will have to re-render because of this. Could we keep it in the `ReportActionsView` and move the `getSortedReportActionsForDisplay` method to `ReportActionsUtils` instead since it is itself utils method.

Then we could get the sorted report actions in the ReportScreen, pass it down to the Report screen components and we can call the `ReportActionsUtils.getSortedReportActionsForDisplay` from the `ReportActionsView` if needed.

Is there anything I am missing?

**Review by mountiny — DISMISSED** [review]
@0xmiroslav How is it looking with your review here?

File: src/pages/home/ReportScreen.js
Comment (gedu) [reply src/pages/home/ReportScreen.js:135]: @mountiny We needed to move up so the `ReportActionsView` and `ReportFooter` receives the same sorted list. 

I could move `getSortedReportActionsForDisplay` into a utils file, but we should keep the sorting at the top. 

About the rerender, it will be the same as before, because when `ReportScreen` needs to rerender when `reportActions` changes, was rerendering both components `ReportActionsView` and `ReportFooter`..., the first one to sort the actions again, and the Footer because of this arrow up feature among others props.

Comment (mountiny) [reply src/pages/home/ReportScreen.js:135]: Thanks for context.

**Comment by mountiny** [comment]
@gedu Can you please resolve the merge conflict?

@0xmiroslav any change you could run through the checklist today, would love to move this one on!

**Review by mountiny — DISMISSED** [review]

File: src/pages/home/report/ReportActionsView.js
Comment (mountiny) [reply src/pages/home/report/ReportActionsView.js:161]: It was removed here https://github.com/Expensify/App/commit/c35193096346ffcfe8bb073c216d314de4419e08

Comment (gedu) [reply src/pages/home/report/ReportActionsView.js:161]: I will look at it thanks

Comment (gedu) [reply src/pages/home/report/ReportActionsView.js:161]: ok, cool, I will update it

File: src/pages/home/ReportScreen.js
Comment (gedu) [reply src/pages/home/ReportScreen.js:130]: Good catch, Yes, for sure, I will try to replicate and add it, thanks

Comment (mountiny) [reply src/pages/home/ReportScreen.js:130]: Not sure if adding the entire props.report is best, there could be some regressions with that I can imagine.

Comment (gedu) [reply src/pages/home/ReportScreen.js:130]: @0xmiroslav Hey, I've been looking at what you wrote and couldn't reproduce it like that, I could, following other steps, and only on web, mobile is working as expected.

https://user-images.githubusercontent.com/1676818/221514798-987ebdad-eb00-4c8e-8dae-ae884da357ac.mp4




> Let's say 2 members (including me) in #announce room!
> -This should not show Split bill on + menu.
On the other hand, looking at the code, I see `#announce` and `#admins` always will show Split Bill on the + menu (if there are 2 members). This was working (above in the video showing the only case isn't working).

Comment (gedu) [reply src/pages/home/ReportScreen.js:130]: I added the check to the reports as you suggested and is fixing the case I found, doing some round of testing and uploading fix, thanks

**Comment by mountiny** [comment]
@0xmiroslav Any chance you could give this another go?

**Comment by gedu** [comment]
> I am still not convinced why we should use `shouldComponentUpdate` in `ReportScreen`. Regarding optimization, for which props change, is it not re-rendered?

@0xmiroslav the page doesn't work as expected

https://user-images.githubusercontent.com/1676818/221616448-dba0ee83-0e32-4849-a929-a6f8983d7979.mov

**Comment by mountiny** [comment]
@0xmiroslav @gedu Personally I would jsut say we want to keep the re-rendering logic same as before hence thats why to re-use the `shouldComponentUpdate` from the ReportActionsView.

Whether we dont need it or the performance gains are not significant idk but I think it should not be scope fo this PR and we can explore that in some other.

Can we focus on wrapping this PR up?

**Comment by mountiny** [comment]
I dont think so for now, unless @0xmiroslav will have some further requests. Would you be able to look into this today? 

I think the shouldComponentUpdate should stay the same to limits chances for regression

**Comment by gedu** [comment]
@0xmiroslav got it, removed the `shouldComponentUpdate`

**Comment by gedu** [comment]
> @gedu can you please re-state the problem if we keep current `shouldComponentUpdate` logic in `ReportActionsView`?

There wasn't an issue keeping `shouldComponentUpdate` in `ReportActionsView`, when I first start the task I think I was missing a bit of context, and lifted it up to have a bit more control, but in being that many cases. We can keep the `shouldComponentUpdate` in `ReportActionsView` and from `ReportScreen` moving the sortedReports into a state so we make sure the components re-renders

File: src/pages/home/ReportScreen.js
Comment (gedu) [reply src/pages/home/ReportScreen.js:138]: True, to make sure we can add it into the state

```
if (!_.isEqual(prevProps.reportActions, this.props.reportActions)) {
            const sortedAndFilteredReportActions = ReportActionsUtils.getSortedReportActionsForDisplay(this.props.reportActions);
            this.setState({sortedAndFilteredReportActions});
        }
 ```

**Comment by gedu** [comment]
@0xmiroslav I can make new videos on each platform if you want, adding `ReportScreen` a state for reports and keeping the `shouldComponentUpdate` in `ReportActionsView`. I just tested and is working

**Comment by mountiny** [comment]
@gedu yep, lets update this

@0xmiroslav thanks for the explanation and taking your time ❤️ I understand now what you meant

**Comment by gedu** [comment]
@0xmiroslav 
<details>
<summary>Web</summary>



https://user-images.githubusercontent.com/1676818/222127883-f735d680-1246-4f06-a59a-0c051115f12d.mp4


https://user-images.githubusercontent.com/1676818/222128062-ba280606-3c24-41b3-bc6c-79546cfa999d.mp4



</details>

<details>
<summary>Mobile Web - Chrome</summary>



https://user-images.githubusercontent.com/1676818/222127931-6dee7e7a-7cfd-4fe4-b783-97065c01371e.mp4



</details>

<details>
<summary>Mobile Web - Safari</summary>



https://user-images.githubusercontent.com/1676818/222127958-c06a8d00-0825-41ff-a8d3-d2efe663bbd0.mp4



</details>

<details>
<summary>Desktop</summary>



https://user-images.githubusercontent.com/1676818/222131650-8281b4cf-646f-4749-bb46-7cf9438c910d.mp4


</details>

<details>
<summary>iOS</summary>



https://user-images.githubusercontent.com/1676818/222128006-39f3653a-5866-4acc-bd88-21e733a68e24.mp4


</details>

<details>
<summary>Android</summary>



https://user-images.githubusercontent.com/1676818/222128027-4a81f660-a8bc-40a0-b71d-fc55cbfa2207.mp4


</details>

**Review by 0xmiroslav — DISMISSED** [review]
@gedu thanks for update. Latest changes look good and test well.
@mountiny All yours!

**Review by mountiny — DISMISSED** [review]
Love where we ended here, thank you both @gedu @0xmiroslav

**Comment by mountiny** [comment]
@0xmiroslav very good point, I have gone ahead and [tested this PR](https://github.com/Expensify/App/pull/15474) and there is a regression. The footer needs all the Closed reportActions in it and the filter method gets rid of them
image

Basically the `getSortedReportActionsForDisplay` method removed the CLOSED report actions as we dont want to display them in the reportActionsView, but we need them in the ReportFooter to show the correct archived report reason. 

cc @gedu

**Comment by mountiny** [comment]
@gedu Could we pass the reportActions untouched to the reportFooter so the closed actions work fine and then just use the  `getSortedReportActionsForDisplay ` on this line as we pass it to the composer https://github.com/Expensify/App/blob/8118c20fd97eb1a253f63f9386e34cf57ab2a652/src/pages/home/report/ReportFooter.js#L106

**Comment by mountiny** [comment]
Or we can try to move [this](https://github.com/Expensify/App/blob/8118c20fd97eb1a253f63f9386e34cf57ab2a652/src/pages/home/report/ReportFooter.js#L73) to the ReportScreen and just pass in the last closed report action which is null by default. But neither solution is ideal imho

**Comment by gedu** [comment]
> For optimization, I suggest to separate this function into sort and filter. So only sort in ReportScreen and then filter in only ReportActionsView. And sort can be reused in getLastClosedReportAction

It was already separated, and in this https://github.com/Expensify/App/commit/c35193096346ffcfe8bb073c216d314de4419e08 commit it was merged

**Comment by gedu** [comment]
> > So only sort in ReportScreen and then filter in only ReportActionsView.
> 
> I am not sure but I feel like we want to filter this also for the Composer to make sure the composer and the view have the identical report action data

What if we add to `ArchivedReportFooter` the `withOnyx({ reportActions: ...})` and only when is an archivedRoom, I just sort and get the closedAction

I mean 
```
const ArchivedReportFooter = (props) => {
    const reportClosedAction = ReportActionsUtils.getLastClosedReportAction(this.props.reportActions);
    const archiveReason = lodashGet(reportClosedAction, 'originalMessage.reason', CONST.REPORT.ARCHIVE_REASON.DEFAULT);
    ...
};

export default compose(
    withLocalize,
    withOnyx({
        reportActions: {
            key: ({route}) => `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${getReportID(route)}`,
            canEvict: false,
        },
        personalDetails: {
            key: ONYXKEYS.PERSONAL_DETAILS,
        },
        policies: {
            key: ONYXKEYS.COLLECTION.POLICY,
        },
    }),
)(ArchivedReportFooter);
```

**Comment by mountiny** [comment]
> main isn't filtering reportActions in ReportFooter right now.
Should this be handled in this PR?

Well I think for the purpose of the composer knowing which message to edit it should know which messages are shown. There should not be really many messages filtered out these days

**Comment by gedu** [comment]
@0xmiroslav and @mountiny pushed

**Review by mountiny — APPROVED** [review]
Thanks I think although its not ideal this works well and its quite clean 
image


tested it works fine in that usecase. Waht do you think @0xmiroslav?

**Comment by mountiny** [comment]
@0xmiroslav this seems like something which will occur in main too, right?

Dont think we should block on this, if it occurs in main, feel free to report it in the channel.

**Review by 0xmiroslav — DISMISSED** [review]
Tested all cases including closed reports.

**Review by mountiny — DISMISSED** [review]
Thank you both!

File: src/pages/home/ReportScreen.js
Comment (luacmartins) [inline src/pages/home/ReportScreen.js:122]: Is this just derived state? It feels like an anti-pattern to store this in state since we can derive it from props. What's the reason for doing this?

File: src/pages/home/report/ReportActionsView.js
Comment (luacmartins) [inline src/pages/home/report/ReportActionsView.js:52]: Do we need to update other usages of reportActions? e.g. [here](https://github.com/Expensify/App/blob/4512136a17062147ce0e38b7029d7769cd903a34/src/components/ReportActionItem/IOUAction.js#L64)

File: src/pages/home/ReportScreen.js
Comment (mountiny) [reply src/pages/home/ReportScreen.js:122]: I see your pointm, I think the idea was to not having to call this function on two places [here](https://github.com/Expensify/App/pull/15207/files#diff-dd4bfa50713397aca8cb40145317936ab0affe19abc2a9c24d6c8849ed75dc9bR287) and [here](https://github.com/Expensify/App/pull/15207/files#diff-dd4bfa50713397aca8cb40145317936ab0affe19abc2a9c24d6c8849ed75dc9bR298)

Comment (gedu) [reply src/pages/home/ReportScreen.js:122]: Also, If the `reportActions` prop changes, we need to re-sort and re-render the component. Even though the sorted data can be derived from props, we still need to store it in state to keep it up-to-date and avoid issues like User B sending a message to User A while User A is in the report, the new message won't show up

**Comment by mountiny** [comment]
@luacmartins comments addressed

File: src/pages/home/report/ReportActionsView.js
Comment (luacmartins) [reply src/pages/home/report/ReportActionsView.js:52]: Cool, thanks for clarifying!

File: src/pages/home/ReportScreen.js
Comment (luacmartins) [reply src/pages/home/ReportScreen.js:122]: I get that, but won't the change in props trigger a re-render of the component anyways (unless we explicitly disabled that in `shouldComponentUpdate`)? Which means we can re-sort the `reportActions` at render time.

Comment (luacmartins) [reply src/pages/home/ReportScreen.js:122]: >I think the idea was to not having to call this function on two places [here](https://github.com/Expensify/App/pull/15207/files#diff-dd4bfa50713397aca8cb40145317936ab0affe19abc2a9c24d6c8849ed75dc9bR287) and [here](https://github.com/Expensify/App/pull/15207/files#diff-dd4bfa50713397aca8cb40145317936ab0affe19abc2a9c24d6c8849ed75dc9bR298)

We could create a variable for it in the render() method

File: src/pages/home/ReportScreen.js
Comment (luacmartins) [reply src/pages/home/ReportScreen.js]: Discussed it [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1677778414008709) and we determined that it's ok to filter/sort in `render()`. I think that we should do that and remove the derived state.

File: src/pages/home/ReportScreen.js
Comment (luacmartins) [reply src/pages/home/ReportScreen.js:122]: We discussed it [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1677778414008709) and determined that we should do the filter/sort in `render()` since performance shouldn't be an issue.

Comment (luacmartins) [reply src/pages/home/ReportScreen.js:122]: Also, as a side note Onyx does support custom props via the `selector` key, as we do [here](https://github.com/Expensify/App/blob/4512136a17062147ce0e38b7029d7769cd903a34/src/pages/home/sidebar/SidebarLinks.js#L290)

Comment (gedu) [reply src/pages/home/ReportScreen.js:122]: @luacmartins Awesome, on my way, thanks for the tips, would be good to add them in `contributingGuides/STYLE.md`

Comment (gedu) [reply src/pages/home/ReportScreen.js:122]: @luacmartins @mountiny  Went through the slack thread and wanna grab some data, is true that there isn't much difference sortingAndFiltering on render, what I could get is a 10% improvement using the derived state.

Screen Shot 2023-03-03 at 12 25 25

Average:
Props: ~300ms
State: ~268ms

Doing Remounting x5 times, and forcing update x15 times

Comment (gedu) [reply src/pages/home/ReportScreen.js:122]: I just saw that I can add a selector to Onyx, I runed the same tests and got almost the same numbers as state:
Screen Shot 2023-03-03 at 12 40 03

The green one

Average:
Onyx with selector: ~260ms

Doing Remounting x5 times, and forcing update x15 times

I would recommend in this cases using the selector from Onyx instead of sorting over the render

All this cases were on a Physical device, OnePlus6 Android 11

Comment (mountiny) [reply src/pages/home/ReportScreen.js:122]: Thanks! Lets use the onyx selector

Comment (luacmartins) [reply src/pages/home/ReportScreen.js:122]: Sounds good. Let me know when the changes are ready for review

Comment (gedu) [reply src/pages/home/ReportScreen.js]: @luacmartins uploaded

**Review by luacmartins — DISMISSED** [review]

**Comment by luacmartins** [comment]
@0xmiroslav would you mind testing again since we had changes?

File: src/pages/home/ReportScreen.js
Comment (mountiny) [reply src/pages/home/ReportScreen.js:193]: Thats correct, I think we should split this into two methods, we need the closed actions in it so I think we should sort it, and then remove the closed actions once we pass that to the ReporActionsView.

**Comment by mountiny** [comment]
@gedu will update this so the footer is using a selector to pull the correct closed message

**Comment by gedu** [comment]
Now `ArchivedReportFooter` using the selector

**Review by luacmartins — APPROVED** [review]
LGTM and tests well. @gedu would you mind adding test steps for the archived footer scenario?

**Comment by gedu** [comment]
> LGTM and tests well. @gedu would you mind adding test steps for the archived footer scenario?

I would paste the steps given to me by @mountiny, could you help me with this? if I miss something

- you would have to create a workspace
- invite two people there
- remove one person from the workspace
- then merge that removed member to the account who is still a member

**Review by mountiny — APPROVED** [review]
Tested well 
image

@luacmartins the archived reason will be confirmed/ tested along with this PR https://github.com/Expensify/App/pull/15474

File: src/pages/home/report/ReportFooter.js
Comment (mountiny) [reply src/pages/home/report/ReportFooter.js:77]: Thanks, created a PR here https://github.com/Expensify/App/pull/15687

File: src/components/ArchivedReportFooter.js
Comment (mountiny) [inline src/components/ArchivedReportFooter.js:97]: This PR caused this regression https://github.com/Expensify/App/issues/16324 because the `getLastClosedReportAction` has been returning `undefined` which in turn did not get swapped for the defaultProps

**Pull request merged**
